verification/src/solidity/compiler_fetcher.rs

```diff
@@ -0,0 +1,331 @@
+#[derive(Default)]
+pub struct CompilerFetcher {
+    releases: HashMap<CompilerVersion, CompilerInfo>,
+    folder: PathBuf,
+}
+
+impl CompilerFetcher {
+    pub async fn new(compilers_list_url: &Url, folder: PathBuf) -> Result<Self, ListError> {
+        Ok(Self {
+            releases: Self::list_releases(compilers_list_url).await?,
+            folder,
+        })
+    }
+
+    async fn list_releases(
+        compilers_list_url: &Url,
+    ) -> Result<HashMap<CompilerVersion, CompilerInfo>, ListError> {
+        let list_json_file: ListJson = reqwest::get(compilers_list_url.to_string())
+            .await
+            .map_err(ListError::ListJsonFetch)?
+            .json()
+            .await
+            .map_err(ListError::ParseListJson)?;
+
+        list_json_file
+            .into_releases(compilers_list_url)
+            .map_err(ListError::Path)
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum FetchError {
+    #[error("version {0} not found")]
+    NotFound(CompilerVersion),
+    #[error("couldn't fetch the file: {0}")]
+    Fetch(reqwest::Error),
+    #[error("couldn't create file: {0}")]
+    File(std::io::Error),
+    #[error("tokio sheduling error: {0}")]
+    Shedule(tokio::task::JoinError),
+}
+
+#[cfg(target_family = "unix")]
+fn create_executable(path: &Path) -> Result<File, std::io::Error> {
+    OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .mode(0o777)
+        .open(path)
+}
+
+#[async_trait]
+impl Fetcher for CompilerFetcher {
+    type Error = FetchError;
+    async fn fetch(&self, ver: &CompilerVersion) -> Result<PathBuf, Self::Error> {
+        let compiler_info = self
+            .releases
+            .get(ver)
+            .ok_or_else(|| FetchError::NotFound(ver.clone()))?;
+
+        let response = reqwest::get(compiler_info.url.clone())
+            .await
+            .map_err(FetchError::Fetch)?;
+        let folder = self.folder.join(ver.to_string());
+        let file = folder.join("solc");
+        let bytes = response.bytes().await.map_err(FetchError::Fetch)?;
+        {
+            let file = file.clone();
+            tokio::task::spawn_blocking(move || -> Result<(), Self::Error> {
+                std::fs::create_dir_all(&folder).map_err(FetchError::File)?;
+                std::fs::remove_file(file.as_path())
+                    .or_else(|e| {
+                        if e.kind() == ErrorKind::NotFound {
+                            Ok(())
+                        } else {
+                            Err(e)
+                        }
+                    })
+                    .map_err(FetchError::File)?;
+                let mut file = create_executable(file.as_path()).map_err(FetchError::File)?;
+                std::io::copy(&mut bytes.as_ref(), &mut file).map_err(FetchError::File)?;
+                Ok(())
+            })
+            .await
+            .map_err(FetchError::Shedule)??;
+        }
+
+        Ok(file)
+    }
+}
+
+mod types {
```

Comment:
What if we name the module `deserialization_types` and move `CompilerInfo` outside? So that we will just separate intermediate types required for deserialization and types used by the CompilerFetcher